add support of gather_object

ZhiyuanChen opened the pull request:
fixes #235

sgugger commented on src/accelerate/utils.py:
```diff
@@ -345,6 +345,40 @@ def gather(tensor):
+    Recursively gather object in a nested list/tuple/dictionary of objects from all devices.
+
+    Args:
+        object (nested list/tuple/dictionary of :obj:`Any`):
```
```suggestion
        object (nested list/tuple/dictionary of any object):
```

sgugger commented on src/accelerate/utils.py:
```diff
@@ -345,6 +345,40 @@ def gather(tensor):
+    def _gpu_gather_object_one(object: Any):
+        output_objects = [None for _ in range(torch.distributed.get_world_size())]
+        torch.distributed.all_gather_object(output_objects, object)
+        return [i for j in output_objects for i in j]
```
I don't understand this line. This presupposes that every element in `output_objects` is iterable, but `object` is supposed to be a primitive object at this point (tensor or other if we remove `error_on_type=True`).

sgugger reviewed (commented):
Thanks a lot for working on this! Could you explain a bit more the inner work of your implementation?

ZhiyuanChen commented:
> I don't understand this line. This presupposes that every element in output_objects is iterable, but object is supposed to be a primitive object at this point (tensor or other if we remove error_on_type=True).

Should we separate `gather_object` to two functions, `gather_iterable` perform `[i for j in output_objects for i in j]` so it works much like original `gather`, while `gather_object` doesn not perform concatenation so it has better support for every objects

sgugger commented:
It can't really concatenate if the tensors are not of the same size, so `gather_object` should not concatenate IMO.

ZhiyuanChen commented:
> It can't really concatenate if the tensors are not of the same size, so `gather_object` should not concatenate IMO.

Yes we can leave it to the user to concat manually.

ZhiyuanChen commented:
I have resolved all conversations, could you please review it again?

sgugger reviewed (approved):
Thanks for addressing all comments, this looks good to merge!

ZhiyuanChen commented:
> Thanks for addressing all comments, this looks good to merge! Could you just run `make style` on your branch to fix the quality failure?

Sorry for this late reply.... I believe I've resolved the quality issue.

sgugger commented on src/accelerate/utils.py:
```diff
@@ -350,6 +350,39 @@ def gather(tensor):
 
+def _gpu_gather_object(object: Any):
+    def _gpu_gather_object_one(object: Any):
+        output_objects = [None for _ in range(state.num_processes)]
```
```suggestion
        output_objects = [None for _ in range(AcceleratorState().num_processes)]
```
`state` is not defined.

sgugger commented on src/accelerate/utils.py:
```diff
@@ -350,6 +350,39 @@ def gather(tensor):
+        torch.distributed.all_gather_object(output_objects, object)
+        return output_objects
+
+    return recursively_apply(_gpu_gather_one, tensor)
```
```suggestion
    return recursively_apply(_gpu_gather_object_one, object)
```

sgugger reviewed (commented):
Here is how to solve the last quality issues (variables not defined).

ZhiyuanChen commented:
Hmmmm, is it possible for us to determine the type of input data and auto call corresponding gather function? I just noticed [torchvison](https://github.com/pytorch/vision/blob/main/references/detection/utils.py#L70-L83) uses `gather_object` at all time

sgugger commented:
That's too magical for Accelerate's philosophy, users should pick which function they want to use.

All green now so we can merge your PR :-)

ZhiyuanChen commented:
> That's too magical for Accelerate's philosophy, users should pick which function they want to use.

Haha, thats the reason why i love this repo so much. Users should be trained programmer and should know what they're doing/supposed to do. 

> All green now so we can merge your PR :-) Thanks again!

Thank you for all the review and comments~
